src/Compilers/Core/Portable/Symbols/SymbolEqualityComparer.cs

```diff
@@ -53,11 +55,11 @@ public bool Equals(ISymbol x, ISymbol y)
             }
             else
             {
-                return x.Equals((object)y);
+                return x.Equals((object?)y);
             }
         }
 
-        public int GetHashCode(ISymbol obj)
+        public int GetHashCode(ISymbol? obj)
```

Comment:
Let's mark this with `[DisallowNull]`. Yes, we handle null, but that's actually a bug. This method should throw an `ArgumentNullException` when `obj` is `null`.

Reply:
See [https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iequalitycomparer-1.gethashcode?view=netframework-4.8](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iequalitycomparer-1.gethashcode?view=netframework-4.8)

---
In reply to: [333635751](https://github.com/dotnet/roslyn/pull/39185#discussion_r333635751) [](ancestors = 333635751)

Reply:
so this has the effect of making the implementation pessimistically think the argument could be null, but the caller will get warned for actually providing a null?

Reply:
I guess I shouldn't overthink it but it feels like this maybe will just break when we get around to #36039

Reply:
> Yes, we handle null, but that's actually a bug.

This is not correct. A derived implementation is allowed to handle null.

Reply:
> I guess I shouldn't overthink it but it feels like this maybe will just break

This case would not break even with attributes affecting implementations. This implementations _widens_ the set of allowed inputs, which is an allowed case (similar to how an implementation can use `string?` for a parameter where the interface only requires `string`).

Reply:
We should follow the BCL guidance on this and not try to widen the contract. We had a very long discussion when reviewing the annotations for IEqualityComparer. Again, yes, we're currently handling `null` by returning a hashcode of 0. This is not correct behavior, `null` does not have a hash code. There are cases when it's acceptable to widen contracts, but this is not one of them.

Reply:
This is correct behavior. It's long been the way to actually properly hash things in maps and support number at the same time. 

Nullable should represent reality and the actual defacto patterns we are using. It shouldn't be lying and pretending the apis are doing something differently.

Reply:
I looked at the implementation of EqualityComparer in CoreCLR this morning. I'm fine with it as is, I must have been misremembering the conversation around the interface.